# Classification Model Evaluation

Evaluate classification predictions using a confusion matrix and common performance metrics such as Accuracy, Recall, Precision, and F1 Score.

In [7]:
import pandas as pd
import numpy as np

from sklearn.metrics import (
    confusion_matrix,
    accuracy_score,
    recall_score,
    precision_score,
    f1_score
)

## 1. Task 1 - Churn Prediction

Convert predicted churn probabilities into class labels using a threshold of 0.5 and evaluate the resulting predictions.

In [8]:
actual = [1, 1, 1, 1, 1, 1, 0, 0, 0, 0]

probability = [
    0.70,
    0.80,
    0.65,
    0.90,
    0.45,
    0.50,
    0.55,
    0.35,
    0.40,
    0.25
]

df_task1 = pd.DataFrame({
    "Actual": actual,
    "Probability": probability
})

threshold = 0.5

df_task1["Prediction"] = np.where(
    df_task1["Probability"] >= threshold,
    1,
    0
)

df_task1

,Actual,Probability,Prediction
0,1,0.70,1
1,1,0.80,1
2,1,0.65,1
3,1,0.90,1
4,1,0.45,0
5,1,0.50,1
6,0,0.55,1
7,0,0.35,0
8,0,0.40,0
9,0,0.25,0


### 1.1 Confusion Matrix

Identify True Positive, True Negative, False Positive, and False Negative predictions.

In [10]:
tp1 = (
    (df_task1["Actual"] == 1)
    & (df_task1["Prediction"] == 1)
).sum()

tn1 = (
    (df_task1["Actual"] == 0)
    & (df_task1["Prediction"] == 0)
).sum()

fp1 = (
    (df_task1["Actual"] == 0)
    & (df_task1["Prediction"] == 1)
).sum()

fn1 = (
    (df_task1["Actual"] == 1)
    & (df_task1["Prediction"] == 0)
).sum()

print(f"True Positive : {tp1}")
print(f"True Negative : {tn1}")
print(f"False Positive: {fp1}")
print(f"False Negative: {fn1}")

True Positive : 5
True Negative : 3
False Positive: 1
False Negative: 1


### 1.2 Classification Metrics

Calculate Accuracy, Recall, Precision, and F1 Score using the confusion matrix values.

In [11]:
accuracy1 = (tp1 + tn1) / (tp1 + tn1 + fp1 + fn1)

recall1 = tp1 / (tp1 + fn1)

precision1 = tp1 / (tp1 + fp1)

f1_1 = 2 * (precision1 * recall1) / (precision1 + recall1)

print(f"Accuracy : {accuracy1:.3f}")
print(f"Recall   : {recall1:.3f}")
print(f"Precision: {precision1:.3f}")
print(f"F1 Score : {f1_1:.3f}")

Accuracy : 0.800
Recall   : 0.833
Precision: 0.833
F1 Score : 0.833


### 1.3 Verification with Scikit-Learn

Verify the manually calculated confusion matrix and classification metrics using Scikit-Learn.

In [13]:
cm1 = confusion_matrix(
    df_task1["Actual"],
    df_task1["Prediction"]
)

accuracy1_sk = accuracy_score(
    df_task1["Actual"],
    df_task1["Prediction"]
)

recall1_sk = recall_score(
    df_task1["Actual"],
    df_task1["Prediction"]
)

precision1_sk = precision_score(
    df_task1["Actual"],
    df_task1["Prediction"]
)

f1_1_sk = f1_score(
    df_task1["Actual"],
    df_task1["Prediction"]
)

print("Confusion Matrix:")
print(cm1)

print(f"\nAccuracy : {accuracy1_sk:.3f}")
print(f"Recall   : {recall1_sk:.3f}")
print(f"Precision: {precision1_sk:.3f}")
print(f"F1 Score : {f1_1_sk:.3f}")

Confusion Matrix:
[[3 1]
 [1 5]]

Accuracy : 0.800
Recall   : 0.833
Precision: 0.833
F1 Score : 0.833


### Task 1 Result

The manually calculated metrics match the Scikit-Learn results, confirming the classification calculations.

## 2. Task 2 - Fraud Detection

Evaluate the fraud detection model using the provided confusion matrix and investigate whether accuracy provides a reliable measure of model performance.

In [14]:
tp2 = 5
fn2 = 5
fp2 = 90
tn2 = 900

In [15]:
accuracy2 = (tp2 + tn2) / (tp2 + tn2 + fp2 + fn2)

recall2 = tp2 / (tp2 + fn2)

precision2 = tp2 / (tp2 + fp2)

f1_2 = 2 * (precision2 * recall2) / (precision2 + recall2)

print(f"Accuracy : {accuracy2:.3f}")
print(f"Recall   : {recall2:.3f}")
print(f"Precision: {precision2:.3f}")
print(f"F1 Score : {f1_2:.3f}")

Accuracy : 0.905
Recall   : 0.500
Precision: 0.053
F1 Score : 0.095


### Task 2 Result

Although the model achieves **90.5% accuracy**, its Recall and Precision scores reveal poor fraud detection performance.

The model detects only half of the fraudulent transactions and produces many false positive predictions. The high Accuracy is therefore misleading due to the strong class imbalance in the dataset.

## 3. Model Evaluation Overview

Compare the two classification scenarios to highlight why model performance should be evaluated using multiple metrics rather than Accuracy alone.

In [16]:
comparison_df = pd.DataFrame({
    "Metric": [
        "Accuracy",
        "Recall",
        "Precision",
        "F1 Score"
    ],
    "Task 1 - Churn": [
        accuracy1,
        recall1,
        precision1,
        f1_1
    ],
    "Task 2 - Fraud": [
        accuracy2,
        recall2,
        precision2,
        f1_2
    ]
})

comparison_df.set_index("Metric").round(3)

,Task 1 - Churn,Task 2 - Fraud
Metric,,
Accuracy,0.800,0.905
Recall,0.833,0.500
Precision,0.833,0.053
F1 Score,0.833,0.095


## Key Findings

- Accuracy provides an overall measure of correct predictions but may be misleading for imbalanced datasets.
- Recall indicates how successfully the model identifies actual positive cases.
- Precision measures how reliable positive predictions are.
- F1 Score provides a balance between Precision and Recall.
- Evaluation metrics should be selected according to the characteristics and business objective of the classification problem.